File: src/utils/paths.ts
// XDG-aware filesystem locations for cache + config.
// macOS doesn't define XDG vars, but the ~/.cache and ~/.config defaults work fine there too.

import { homedir } from "node:os";
import { join } from "node:path";

const APP = "manga-cli";

function baseDir(xdgVar: string, fallback: string): string {
  const xdg = process.env[xdgVar];
  if (xdg && xdg.startsWith("/")) return xdg;
  return join(homedir(), fallback);
}

export const CACHE_DIR = join(baseDir("XDG_CACHE_HOME", ".cache"), APP);
export const CONFIG_DIR = join(baseDir("XDG_CONFIG_HOME", ".config"), APP);

export const CONFIG_FILE = join(CONFIG_DIR, "config.json");
export const HISTORY_FILE = join(CONFIG_DIR, "history.json");
export const FOLLOWS_FILE = join(CONFIG_DIR, "follows.json");

export const COVERS_DIR = join(CACHE_DIR, "covers");
export const PAGES_DIR = join(CACHE_DIR, "pages");
export const SEARCH_CACHE = join(CACHE_DIR, "search");
export const MANGA_CACHE = join(CACHE_DIR, "manga");
export const CHAPTERS_CACHE = join(CACHE_DIR, "chapters");

/** Expand a leading ~ to the user's home directory. */
export function expandTilde(p: string): string {
  if (p === "~") return homedir();
  if (p.startsWith("~/")) return join(homedir(), p.slice(2));
  return p;
}

File: go-rewrite/internal/util/paths.go
// Package util: XDG-aware filesystem locations, config, and the tiny disk cache.
// File formats and paths are byte-compatible with the TypeScript implementation,
// so both binaries share config, history, follows, and caches.
package util

import (
	"os"
	"path/filepath"
	"strings"
)

const app = "manga-cli"

func baseDir(xdgVar, fallback string) string {
	if xdg := os.Getenv(xdgVar); strings.HasPrefix(xdg, "/") {
		return xdg
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, fallback)
}

var (
	CacheDir  = filepath.Join(baseDir("XDG_CACHE_HOME", ".cache"), app)
	ConfigDir = filepath.Join(baseDir("XDG_CONFIG_HOME", ".config"), app)

	ConfigFile  = filepath.Join(ConfigDir, "config.json")
	HistoryFile = filepath.Join(ConfigDir, "history.json")
	FollowsFile = filepath.Join(ConfigDir, "follows.json")

	CoversDir     = filepath.Join(CacheDir, "covers")
	PagesDir      = filepath.Join(CacheDir, "pages")
	SearchCache   = filepath.Join(CacheDir, "search")
	MangaCache    = filepath.Join(CacheDir, "manga")
	ChaptersCache = filepath.Join(CacheDir, "chapters")
)

// ExpandTilde expands a leading ~ to the user's home directory.
func ExpandTilde(p string) string {
	home, _ := os.UserHomeDir()
	if p == "~" {
		return home
	}
	if strings.HasPrefix(p, "~/") {
		return filepath.Join(home, p[2:])
	}
	return p
}
